### PR TITLE
(SDI-2181) document new plugin authoring and publishing

### PR DESCRIPTION
**Pull request by nanliu**

### CODE REVIEW

File: docs/PLUGIN_AUTHORING.md [hunk 1]
```diff
@@ -17,152 +17,134 @@ See the License for the specific language governing permissions and
...
+    * [Plugin Release](#plugin-release)
+    * [Plugin Metadata](#plugin-metadata)
+    * [Documentation](#documentation)
 
-Bon Appétit! :stew:
+## Overview
 
-## Plugin Authoring
-Snap itself runs as a master daemon with the core functionality that may load and unload plugin processes via either CLI or HTTP APIs.
+Snap daemon runs as a service which provides the core functionality such as plugin management and task scheduling. Snap plugins extend Snap by providing additional capability to gather, transform, or publish metrics. Since plugins are standalone programs, they can be written in any language.
```
Comment (IRCody) [inline docs/PLUGIN_AUTHORING.md]: Instead of any language (even though that's technically correct) I'd probably limit it to languages supported by gRPC.

File: docs/PLUGIN_AUTHORING.md [hunk 2]
```diff
@@ -17,152 +17,134 @@ See the License for the specific language governing permissions and
...
+* [snap-plugin-lib-go](https://github.com/intelsdi-x/snap-plugin-lib-go)
+* [snap-plugin-lib-cpp](https://github.com/intelsdi-x/snap-plugin-lib-cpp)
+* [snap-plugin-lib-py](https://github.com/intelsdi-x/snap-plugin-lib-py)
 
-Before starting writing Snap plugins, check out the [Plugin Catalog](https://github.com/intelsdi-x/snap/blob/master/docs/PLUGIN_CATALOG.md) to see if any suit your needs. If not, you need to reference the plugin packages that defines the type of structures and interfaces inside snap and then write plugin endpoints to implement the defined interfaces.
+Communication between Snap daemon and plugins uses gRPC. So even if a plugin library isn't available in the language of your choice, you can still write a plugin using [gRPC library](http://grpc.io/docs) as a starting point. However this requires additional knowledge about Snap API, and is beyond the scope of this document.
```
Comment (IRCody) [inline docs/PLUGIN_AUTHORING.md]: Should we also add gRPC/protobuf to the "requires additional knowledge" section?

Some small things.
`uses` -> `use`
`using gRPC library` -> `using the gRPC library`

File: docs/PLUGIN_AUTHORING.md [hunk 3]
```diff
@@ -17,152 +17,134 @@ See the License for the specific language governing permissions and
...
+* `snap-plugin-publisher-influxdb`
 
+### Plugin Metric Namespace
 
-Snap validates the metrics exposed by the plugin and, if validation fails, an error is returned and the plugin is not loaded.
+When gathering data in collector plugins, each metric requires a namespace and a description of what data is being gathered. The metric namespace should contain the org, name of plugin, and the metric's name:
 
-##### c) static and dynamic metrics
-Snap supports both static and dynamic metrics.  You can find more detail about static and dynamic metrics [here](./METRICS.md).
+`/[organization]/[plugin name]/[plugin namespace]/[metric name]`
```
Comment (IRCody) [inline docs/PLUGIN_AUTHORING.md]: I think we should specify that <metric name> can be multiple sections. This kind've reads like it's always 4.

File: docs/PLUGIN_AUTHORING.md [hunk 4]
```diff
@@ -17,152 +17,134 @@ See the License for the specific language governing permissions and
...
-    - other punctuations: `. , ; ? !`
+* collector: gathering metrics
+* processor: transforming metrics
+* publisher: publishing metrics
 
-##### b) a wildcard in the end
+### Plugin Name
 
-Example:
+The plugin name should follow this format: `snap-plugin-[type]-[name]`
```
Comment (IRCody) [inline docs/PLUGIN_AUTHORING.md]: Maybe rephrase this to be a convention rather than a requirement?

File: docs/PLUGIN_AUTHORING.md [hunk 5]
```diff
@@ -17,152 +17,144 @@ See the License for the specific language governing permissions and
...
 
-A Snap plugin is a program, or a set of functions or services, written in Go or any language; that may seamlessly integrate with snap as executables. 
+The following libraries are available to simplify the process of writing a plugin:
 
-Communication between Snap and plugins uses RPC either through HTTP or TCP protocols. HTTP gRPC is good for any language to use that gRPC supports (see [gRPC docs](http://grpc.io/docs)) while the native client is only suitable for plugins written in Golang. When a plugin is written using one of the available snap-plugin-libs ([snap-plugin-lib-go](https://github.com/intelsdi-x/snap-plugin-lib-go), [snap-plugin-lib-py](https://github.com/intelsdi-x/snap-plugin-lib-py), or [snap-plugin-lib-cpp](https://github.com/intelsdi-x/snap-plugin-lib-cpp)) Google Protobuf, a binary serialization format, is used to encode/decode the data (see [plugin.proto](https://github.com/intelsdi-x/snap/blob/master/control/plugin/rpc/plugin.proto)). 
+* [snap-plugin-lib-go](https://github.com/intelsdi-x/snap-plugin-lib-go)
```
Comment (kindermoumoute) [inline docs/PLUGIN_AUTHORING.md]: Is it possible to link to anchors of the readme title? (e.g. [link](https://github.com/intelsdi-x/snap-plugin-lib-go#snap-plugin-library-for-go)

File: docs/PLUGIN_AUTHORING.md [hunk 6]
```diff
@@ -17,152 +17,144 @@ See the License for the specific language governing permissions and
...
+Communication between Snap daemon and plugins use gRPC. So even if a plugin library isn't available in the language of your choice, you can still write a plugin using the [gRPC library](http://grpc.io/docs) as a starting point. However this requires additional knowledge about Snap API, and is beyond the scope of this document.
 
-### Plugin Naming, Files, and Directory    
-Snap supports three type of plugins. They are collectors, processors, and publishers.  The plugin project name should use the following format:  
->snap-plugin-[type]-[name]
+A few notes before we get started:
+* Before writing a new Snap plugin, please check out the [Plugin Catalog](./PLUGIN_CATALOG.md) to see if any existing plugins meet your needs.
+* If you need any assistance, please reach out on [slack #snap-developers channel](https://intelsdi-x.herokuapp.com/).
```
Comment (kindermoumoute) [inline docs/PLUGIN_AUTHORING.md]: Slack*

File: docs/PLUGIN_AUTHORING.md [hunk 7]
```diff
@@ -17,152 +17,144 @@ See the License for the specific language governing permissions and
...
+* `/acme/water/NewYork/usage`
+* `/acme/water/NewYork/rainfall`
+* `/acme/water/NewYork/Bronx/usage`
+* `/acme/water/NewYork/Chelsea/usage`
+* `/acme/water/NewYork/Willisamburg/usage`
+* `/acme/water/Portland/usage`
+* `/acme/water/Portland/rainfall`
+* `/acme/water/Portland/Hollywood/usage`
+* `/acme/water/Portland/Pearl/usage`
+* `/acme/water/Portland/StJohn/usage`
```
Comment (kindermoumoute) [inline docs/PLUGIN_AUTHORING.md]: Is it a good place to talk about dynamic metrics?

File: docs/PLUGIN_AUTHORING.md [hunk 8]
```diff
@@ -17,152 +17,144 @@ See the License for the specific language governing permissions and
...
 
-Building main.go generates a binary executable. You may choose to sign the executable with our [plugin signing](https://github.com/intelsdi-x/snap/blob/master/docs/PLUGIN_SIGNING.md).
+A plugin can have any number of internal plugin namespace. It's up to the plugin author on how to organize the metrics in a meaningful way. This information should be documented in the README for ease of usage.
 
-### Localization
-All comments and READMEs within the plugin code should be in English.  For different languages, include appropriate translation files within the plugin package for internationalization.
+The metric namespace cannot contain [Unicode character U+1F422](http://www.fileformat.info/info/unicode/char/1f422/index.htm) i.e. :turtle:
```
Comment (kindermoumoute) [inline docs/PLUGIN_AUTHORING.md]: I would remove this line.
1. Actually it can contain it. 
2. This is "kinda" an easter egg.

File: docs/PLUGIN_AUTHORING.md (hunk 2 above)
Comment (nanliu) [inline docs/PLUGIN_AUTHORING.md]: Should we have an advanced section? Anything you recommend preserving and copying over?

File: docs/PLUGIN_AUTHORING.md (hunk 3 above)
Comment (nanliu) [inline docs/PLUGIN_AUTHORING.md]: Updated with improved examples. Let me know if this is sufficient.

File: docs/PLUGIN_AUTHORING.md [hunk 9]
```diff
@@ -17,152 +17,144 @@ See the License for the specific language governing permissions and
...
-Snap provides the encryption capability for both HTTP and TCP clients. The communication between the Snap daemon and the plugins is encrypted by default. Should you want to disable the encrypted communication, when authoring a plugin, use the `Unsecure` option for your plugin's meta:
-```
-//Meta returns the metadata for MyPlugin
-func Meta() *plugin.PluginMeta {
-    return plugin.NewPluginMeta(name, ver, type, ct, ct2, plugin.Unsecure(true))
+NOTE: The intelsdi-x project reserves the org namespace `/intel`.
+
+### Plugin Interface
+
+Depending on the type of plugin, they must implement several methods to satisfy the appropriate interfaces. Please see the plugin library for language specific examples and documentation.
```
Comment (kindermoumoute) [inline docs/PLUGIN_AUTHORING.md]: Is it possible to create a title under "overview" called "plugin libraries", and you can link to this title from this sentence. Or is it too much?

File: docs/PLUGIN_AUTHORING.md (hunk 3 above)
Comment (IRCody) [inline docs/PLUGIN_AUTHORING.md]: Yea I like the new examples.

File: docs/PLUGIN_AUTHORING.md (hunk 5 above)
Comment (nanliu) [inline docs/PLUGIN_AUTHORING.md]: It doesn't exist for the other plugins, and it will be brittle since it depends on the H1 string for the readme file.

**Comment by mjbrender** [comment]
Just to be clear, we've deleted the PLUGIN_BEST_PRACTICES.md in this PR @intelsdi-x/snap-maintainers. The file was out of date compared to our current best practices, but may be want to be recreated in the future (in this repository or in the plugin libraries directly). Adding this comment as an FYI.
